Demo app: authenticating a script (#701)
inrupt/solid-client-authn-js: packages/node/example/bootstrappedApp/README.md

# Demo: authenticated a NodeJS script

This demo shows how you can run an authenticated script to interact with your Pod.
Check out the code in `src/app.js` for more details.

Note that until [this issue](https://github.com/solid/node-solid-server/issues/1533)
is resolved, this does not work against NSS, and only works against a compliant
OIDC implementation, such as Inrupt's ESS (e.g. deployed at `https://pod.inrupt.com`).

## Installing the app

> To use the app, first run `npm ci` in the root repository.

Reviewer: Seeing how it imports from `../../../dist`, I presume you'd also need to run `npm run build` or something in the root?

Author: Good point, the instructions weren't clear on this. Running `npm ci` at the root of the repo will trigger the build, but then you'd still have to `npm ci` the demo app itself. I clarified this in 3738e8c9f8c559d8d371ff59f0d6642cbd27e6fa.

Reviewer: Ah btw, about that: do we want this to fail for people following instructions if we forget to update the lockfile? (i.e. is it really better for people to run `npm ci` rather than `npm install`?) Is it likely that it will cause inexplicable errors when running this if not?

> # Demo: authenticated a NodeJS script

Reviewer 2: ```suggestion
# Demo: an authenticated NodeJS script
```

This demo shows how you can run an authenticated script to interact with your Pod.
Check out the code in `src/app.js` for more details.

Note that until [this issue](https://github.com/solid/node-solid-server/issues/1533)

> is resolved, this does not work against NSS, and only works against a compliant

Reviewer 2: ```suggestion
is resolved, this will not work against NSS. This demo is only intended to work against a compliant
```

OIDC implementation, such as Inrupt's ESS (e.g. deployed at `https://pod.inrupt.com`).

## Installing the app

> To use the app, first run `npm ci` at the root of the repository: it will trigger

Reviewer 2: ```suggestion
To properly install this demo app, first run `npm ci` at the root of the repository, which will trigger
```

the build of the library. Then, run `npm ci` again, this time in
`packages/node/example/bootstrappedApp/`. You should be all set.

## Getting a token

To get all the credentials needed to authenticate your script, you can `cd src/`,
and then run `node bootstrap.js --oidcProvider <your identity provider>`. 

> This utility script will show in the console an IRI in the domain of your favorite

Reviewer 2: ```suggestion
This utility script will output to the console an IRI in the domain of your specified
```

identity provider (e.g. `https://broker.pod.inrupt.com/`) that you should open

> in a web browser. There, you should be able to log in your identity provider. Once

Reviewer 2: ```suggestion
in a web browser. There, you should be able to log into your identity provider. Once
```

logged in, you should be redirected to a page that tells you the required information
has been sent to the boostrap app. If you go back to the terminal, you'll see

> your client ID, secret and refresh token. Save those.

Reviewer 2: ```suggestion
your client ID, client secret and refresh token. Save these values for use in the next step.

## Using the token

> All you have to do now is to use the credentials previously obtained to get an

Reviewer 2: ```suggestion
All you need to do now is to use the credentials obtained from the previous step to get an
```

> access token. `app.js` shows a minimal app issuing an authenticated request. It

Reviewer 2: Should this source file be named something a bit more descriptive, e.g. `authenticatedScript.js` for example...?

Author: Fixed in f8b6c127da465222224fdccc7dc07ca254567302

> expects a couple of arguments: 

Reviewer 2: ```suggestion
expects a number of arguments: 
```

```

> node app.js --id <the client id> --secret <the client secret>  --token <the refresh token> --issuer <the issuer that issued the token> --resource <the private resource you want to access>

Reviewer 2: I'd use more descriptive command-line args, i.e. `clientId`, `clientSecret`, `accessToken`, `identityProvider`.

Author: Fixed in f8b6c127da465222224fdccc7dc07ca254567302
